src/graph_world/generators/cabam_simulator.py

```diff
@@ -0,0 +1,227 @@
+
+      !!!Identical and equal to all _task_Dataset classes!!!
+
+      graph: graph-tool Graph object.
+      graph_memberships: list of integer node classes.
+      node_features: numpy array of node features.
+      feature_memberships: list of integer node feature classes.
+      edge_features: map from edge tuple to numpy array. Only stores undirected
+        edges, i.e. (0, 1) will be in the map, but (1, 0) will not be.
+    """
+    graph: graph_tool.Graph = Ellipsis # ok
+    graph_memberships: np.ndarray = Ellipsis # we have - node labels
+    node_features: np.ndarray = Ellipsis # node features are also our labels
+    feature_memberships: np.ndarray = Ellipsis # index i contains feature group of node i so this is also node labels
+    edge_features: Dict[Tuple[int, int], np.ndarray] = Ellipsis # edge set but what is the map?
+    # do i even need these or is it enough to have just graph
+    # i think they are used for paramterising certain parts of SBM generation - where my case only requires 
+    # a singular generation fn - no nevernind they are used also in node classification utils
+    
+def nxToGraphWorldData(G, node_labels, cabam_data):
+  cabam_data.graph_memberships = list(node_labels) # Memberships is integer node class list
+  cabam_data.node_features = np.array(sorted(list(set(node_labels)))) # Node features are set of possible labels
+  cabam_data.feature_memberships = list(node_labels) # which feature does each node belong to? features are labels here so this is also node label class list
+
+  # Manipulate G into cabam_data.graph gt object
+  nx_edges = list(G.edges())
+  nx_nodes = list(G.nodes())
+  cabam_data.graph = graph_tool.Graph(directed=False)
+  
+  # Add the nodes
+  vertices = {} # vertex mapping for tracking edges later
+  for node in nx_nodes:
+      # Create the vertex and annotate for our edges later
+      v = cabam_data.graph.add_vertex()
+      vertices[node] = v
+  # Add the edges
+  for src, dst in nx_edges:
+      # Look up the vertex structs from our vertices mapping and add edge.
+      e = cabam_data.graph.add_edge(vertices[src], vertices[dst])
+
+  # Add null edge features
+  cabam_data.edge_features = {}
+  for edge in cabam_data.graph.edges():
+    vertex1 = int(edge.source())
+    vertex2 = int(edge.target())
+    edge_tuple = tuple(sorted((vertex1, vertex2)))
+    cabam_data.edge_features[edge_tuple] = 0
+  return cabam_data
+
+def GenerateAssortativityDict(p_in, p_out):
+    return {0: p_in, 1:p_out}
+
+def GenerateCABAMGraph(n, m, p_in, p_out):
+  result = CABAM()
+  G, node_labels = cabam_graph_generation(n=n, m=m, c_probs=GenerateAssortativityDict(p_in,p_out) )
+  nxToGraphWorldData(G, node_labels, result)
+  return result
+
+def cabam_graph_generation(n, m, c=2, native_probs=[0.5, 0.5], c_probs={1: 0.5, 0: 0.5}, logger=None):
+    '''
+    Main function for CABAM graph generation. Taken directly from https://github.com/nshah171/cabam-graph-generation/blob/master/cabam_utils.py
```

Comment:
@bperozzi Is there a better way to credit Neil and this code?

Reply:
@muzz-yasir I checked with Bryan, and a better way to go about this is as follows:

1. Modify our [Docker file](https://github.com/google-research/graphworld/blob/main/Dockerfile) with a RUN statement that clones Neil's CABAM repository. Prior to the copy call at line 69 of the current Docker file, make sure the repository files are in a location in /src/ such that the CABAM functions can be imported.
2. Instead copying and adding the CABAM code directly, add code to import the function from the downloaded repo files.